dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 Dataline
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.dataline.workers.singer;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import com.google.common.collect.Lists;
+import io.dataline.config.SingerCatalog;
+import io.dataline.config.SingerColumn;
+import io.dataline.config.SingerColumnMap;
+import io.dataline.config.SingerMessage;
+import io.dataline.config.SingerStream;
+import io.dataline.config.SingerTableSchema;
+import io.dataline.config.SingerType;
+import io.dataline.config.StandardDiscoverSchemaInput;
+import io.dataline.config.StandardTapConfig;
+import io.dataline.workers.DefaultSyncWorker;
+import io.dataline.workers.InvalidCredentialsException;
+import io.dataline.workers.JobStatus;
+import io.dataline.workers.OutputAndStatus;
+import io.dataline.workers.StreamFactory;
+import io.dataline.workers.TestConfigHelpers;
+import io.dataline.workers.WorkerUtils;
+import io.dataline.workers.process.ProcessBuilderFactory;
+import io.dataline.workers.protocol.singer.MessageUtils;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class SingerTapFactoryTest {
+  private static final String IMAGE_NAME = "hudi:latest";
+  private static final String JOB_ROOT_PREFIX = "workspace";
+  private static final String DISCOVER_SCHEMA_JOB_ROOT_PREFIX = "discover";
```

Comment:
Why do you need to redefine these statics?

Reply:
:+1: used the static that's defined in the main class for discover. the job root for this class is just any string from its point of view, so leaving that one.